core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -403,32 +403,47 @@ class DAGScheduler(
     parents.toList
   }
 
-  /** Find ancestor shuffle dependencies that are not registered in shuffleToMapStage yet */
-  private def getAncestorShuffleDependencies(rdd: RDD[_]): Stack[ShuffleDependency[_, _, _]] = {
-    val parents = new Stack[ShuffleDependency[_, _, _]]
+  /**
+   * Find ancestor shuffle dependencies that are not registered in shuffleToMapStage yet.
+   * This is done in topological order to create ancestor stages first to ensure that the result
+   * stage graph is correctly built.
+   */
+  private def getAncestorShuffleDependencies(rdd: RDD[_]): Seq[ShuffleDependency[_, _, _]] = {
+    val parents = new ArrayBuffer[ShuffleDependency[_, _, _]]
     val visited = new HashSet[RDD[_]]
     // We are manually maintaining a stack here to prevent StackOverflowError
     // caused by recursively visiting
     val waitingForVisit = new Stack[RDD[_]]
     def visit(r: RDD[_]) {
-      if (!visited(r)) {
-        visited += r
-        for (dep <- r.dependencies) {
-          dep match {
-            case shufDep: ShuffleDependency[_, _, _] =>
-              if (!shuffleToMapStage.contains(shufDep.shuffleId)) {
-                parents.push(shufDep)
-              }
-            case _ =>
+      if (visited(r)) {
+        waitingForVisit.pop()
+      } else {
+        val visitedShuffleDeps = new ArrayBuffer[ShuffleDependency[_, _, _]]
+        val unvisitedDeps = new ArrayBuffer[Dependency[_]]
+
+        r.dependencies.foreach {
+          case dep: ShuffleDependency[_, _, _] if !shuffleToMapStage.contains(dep.shuffleId) =>
+            if (visited(dep.rdd)) visitedShuffleDeps += dep
+            else unvisitedDeps += dep
+          case dep if !visited(dep.rdd) => unvisitedDeps += dep
```

Comment:
Can you put the "unvisitedDeps += dep" on the next line (and indented)? AFAIK that's more consistent with the style used elsewhere in Spark